File: frontend/src/features/production/domain/procedurePrices.ts
export type ProcedurePriceItem = {
  procedure_id: number
  procedure_name: string
  unit_price: number | string | null
  referenced: boolean
}

export type ProcedurePriceScope = {
  row_type: 'formal'
  product_id: number
  product_version: number
  origin_flow_node_id: string
  flow_node_id: string
  product_name: string
  factory_code: string
  product_bom_id: number | null
  workshop_id: number
  workshop_name: string
  part_name: string
  part_no: string
  confirmed: boolean
  can_cancel: boolean
  confirmed_at: string | null
  confirmed_by: string | null
  procedures: ProcedurePriceItem[]
}

export type TemporaryWorkOrderPriceItem = {
  row_type: 'temporary'
  work_order_id: number
  work_order_no: string
  product_id: number
  product_version: number
  origin_flow_node_id: string
  flow_node_id: string
  product_name: string
  factory_code: string
  part_name: string
  part_no: string
  workshop_name: string
  procedure_name: string
  unit_price: number | string | null
  status: 'open' | 'closed' | 'cancelled'
  created_at: string
}

export type ProcedurePriceListItem = ProcedurePriceScope | TemporaryWorkOrderPriceItem

export type ProcedurePriceRevision = {
  id: number
  target_type: 'formal' | 'temporary'
  target_label: string
  previous_unit_price: number | string | null
  new_unit_price: number | string | null
  actor_username: string
  created_at: string
}

File: backend/src/schemas/procedure_prices.py
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class ProcedurePriceModel(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)


class ProcedurePriceItem(ProcedurePriceModel):
    procedure_id: int
    procedure_name: str
    unit_price: Decimal | None
    referenced: bool


class ProcedurePriceScope(ProcedurePriceModel):
    product_id: int
    product_version: int
    product_name: str
    factory_code: str
    product_bom_id: int | None
    origin_flow_node_id: str
    flow_node_id: str
    workshop_id: int
    workshop_name: str
    part_name: str
    part_no: str
    procedures: list[ProcedurePriceItem]


class ProcedurePriceListEnvelope(ProcedurePriceModel):
    data: list[ProcedurePriceScope]
    total: int


class ProcedurePriceInput(ProcedurePriceModel):
    procedure_id: int | None = Field(default=None, gt=0)
    procedure_name: str = Field(min_length=1, max_length=200)
    unit_price: Decimal | None = Field(default=None, ge=0, max_digits=12, decimal_places=2)


class ProcedurePriceUpdate(ProcedurePriceModel):
    procedures: list[ProcedurePriceInput] = Field(default_factory=list, max_length=100)
